File: api_contracts.py
"""Runtime API contract validation for SmartNotes HTTP endpoints."""

API_SCHEMA_VERSION = "v1"


class ContractValidationError(Exception):
    def __init__(self, message, details=None):
        super().__init__(message)
        self.message = message
        self.details = details or {}


def _require_object(payload, field_name="body"):
    if not isinstance(payload, dict):
        raise ContractValidationError(f"Invalid {field_name}: expected JSON object")


def _require_non_empty_string(payload, key):
    value = payload.get(key)
    if not isinstance(value, str) or not value.strip():
        raise ContractValidationError(f"Missing required field: {key}")
    return value


def _require_optional_bool(payload, key):
    if key in payload and not isinstance(payload.get(key), bool):
        raise ContractValidationError(f"Invalid field type: {key} must be boolean")


def _validate_messages(payload):
    messages = payload.get("messages")
    if not isinstance(messages, list) or len(messages) == 0:
        raise ContractValidationError("Missing required field: messages[]")
    for idx, item in enumerate(messages):
        if not isinstance(item, dict):
            raise ContractValidationError(f"Invalid messages[{idx}]: expected object")
        role = item.get("role")
        if not isinstance(role, str) or not role.strip():
            raise ContractValidationError(f"Invalid messages[{idx}].role")
        if "content" not in item:
            raise ContractValidationError(f"Missing messages[{idx}].content")


def _validate_notes(payload):
    notes = payload.get("notes")
    if not isinstance(notes, list):
        raise ContractValidationError("Missing required field: notes[]")


def validate_api_request(method, path, payload):
    """Validate incoming request payload for known API endpoints."""
    if method != "POST":
        return payload

    _require_object(payload)

    if path == "/api/generate":
        _require_non_empty_string(payload, "prompt")
        _require_optional_bool(payload, "stream")
        return payload

    if path == "/api/chat":
        _validate_messages(payload)
        _require_optional_bool(payload, "stream")
        return payload

    if path == "/api/embeddings":
        prompt = payload.get("prompt") or payload.get("input")
        if not isinstance(prompt, str) or not prompt.strip():
            raise ContractValidationError("Missing required field: prompt")
        return payload

    if path == "/api/model":
        _require_non_empty_string(payload, "model")
        return payload

    if path == "/api/file-notes/sync":
        _validate_notes(payload)
        return payload

    return payload


def validate_api_response(path, payload, status=200):
    """Validate known response envelopes for local API handlers."""
    if not isinstance(payload, dict):
        return

    if status >= 400:
        if "error" not in payload:
            raise ContractValidationError("Error responses must include 'error'")
        return

    if path == "/api/default-model":
        model = payload.get("model")
        if not isinstance(model, str) or not model.strip():
            raise ContractValidationError("Invalid response: model")
        if not isinstance(payload.get("source"), str):
            raise ContractValidationError("Invalid response: source")
        return

    if path == "/api/model":
        if not isinstance(payload.get("ok"), bool):
            raise ContractValidationError("Invalid response: ok")
        model = payload.get("model")
        if not isinstance(model, str) or not model.strip():
            raise ContractValidationError("Invalid response: model")
        return

    if path == "/api/file-notes/load":
        if not isinstance(payload.get("notes"), list):
            raise ContractValidationError("Invalid response: notes")
        if not isinstance(payload.get("count"), int):
            raise ContractValidationError("Invalid response: count")
        return

    if path == "/api/file-notes/sync":
        if not isinstance(payload.get("ok"), bool):
            raise ContractValidationError("Invalid response: ok")
        if not isinstance(payload.get("written"), int):
            raise ContractValidationError("Invalid response: written")
        if not isinstance(payload.get("files"), list):
            raise ContractValidationError("Invalid response: files")

File: api-contracts.js
(function (global) {
    const API_SCHEMA_VERSION = 'v1';

    function createContractError(message) {
        const err = new Error(message);
        err.name = 'ApiContractError';
        return err;
    }

    function ensureObject(value, context) {
        if (!value || typeof value !== 'object' || Array.isArray(value)) {
            throw createContractError(`Invalid ${context}: expected object`);
        }
    }

    function ensureNonEmptyString(value, field) {
        if (typeof value !== 'string' || !value.trim()) {
            throw createContractError(`Invalid field: ${field}`);
        }
    }

    function validateRequest(path, payload) {
        ensureObject(payload, 'request payload');

        if (path === '/api/generate') {
            ensureNonEmptyString(payload.prompt, 'prompt');
            if (Object.prototype.hasOwnProperty.call(payload, 'stream') && typeof payload.stream !== 'boolean') {
                throw createContractError('Invalid field: stream must be boolean');
            }
            return payload;
        }

        if (path === '/api/chat') {
            if (!Array.isArray(payload.messages) || payload.messages.length === 0) {
                throw createContractError('Invalid field: messages must be a non-empty array');
            }
            return payload;
        }

        if (path === '/api/embeddings') {
            const prompt = payload.prompt || payload.input;
            ensureNonEmptyString(prompt, 'prompt');
            return payload;
        }

        if (path === '/api/model') {
            ensureNonEmptyString(payload.model, 'model');
            return payload;
        }

        if (path === '/api/file-notes/sync') {
            if (!Array.isArray(payload.notes)) {
                throw createContractError('Invalid field: notes must be an array');
            }
            return payload;
        }

        return payload;
    }

    function validateResponse(path, payload) {
        ensureObject(payload, 'response payload');

        if (payload.error && typeof payload.error !== 'string') {
            throw createContractError('Invalid response: error must be a string');
        }

        if (path === '/api/default-model') {
            ensureNonEmptyString(payload.model, 'model');
            if (payload.schemaVersion && payload.schemaVersion !== API_SCHEMA_VERSION) {
                throw createContractError(`Unsupported schemaVersion: ${payload.schemaVersion}`);
            }
            return payload;
        }

        if (path === '/api/model') {
            if (typeof payload.ok !== 'boolean') {
                throw createContractError('Invalid response: ok must be boolean');
            }
            ensureNonEmptyString(payload.model, 'model');
            return payload;
        }

        if (path === '/api/chat') {
            if (!payload.message && !payload.response && !payload.error) {
                throw createContractError('Invalid response: missing chat content');
            }
            return payload;
        }

        if (path === '/api/generate') {
            if (!Object.prototype.hasOwnProperty.call(payload, 'response') && !payload.error) {
                throw createContractError('Invalid response: missing response field');
            }
            return payload;
        }

        if (path === '/api/embeddings') {
            if (!Array.isArray(payload.embedding) && !payload.error) {
                throw createContractError('Invalid response: embedding must be an array');
            }
            return payload;
        }

        return payload;
    }

    const api = {
        API_SCHEMA_VERSION,
        validateRequest,
        validateResponse,
        createContractError
    };

    if (typeof module !== 'undefined' && module.exports) {
        module.exports = api;
    }

    global.ApiContracts = api;
})(typeof window !== 'undefined' ? window : globalThis);
